### PR TITLE
Switch to OpenAstronomy GitHub Actions workflow to build and publish wheels

### PR DESCRIPTION
**Pull request by astrofrog**
### Description

This switches the wheel building and publishing machinery from Azure to GitHub Actions, making use of the OpenAstronomy workflow (https://github.com/OpenAstronomy/github-actions-workflows).

To start with, I am copying over the configuration as-is, and we can then see if any improvements can be made. Remaining things to do:

* [ ] <s>Check if we can enable the musllinux wheels</s> - still some issues, will investigate after this PR
* [x] Set up nightly build machinery, which will require changing the URL for development wheels to a different URL (assuming we do this, we should clearly warn people in advance - nothing will break since the old dev wheels will still be available at the old URL but no new dev wheels will be added there)
* [x] Consider whether to drop 32-bit wheels for Linux - I think for now we should actually leave it in this PR so that we can backport it to 5.0.x and 5.1.x and then remove 32-bit Linux support in ``main`` (if we decide to)
* [x] Don't build all the wheels for all PRs - build only a couple for regular PRs, and all on release branches and tags.
* [x] Add back all the regular CI builds (remove the TMP commit)

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by astrofrog** [comment]
The config in this PR currently will only publish dev wheels, to anaconda.org. This is different from the Azure PyPI URL that we used to upload to. I think what we should do is go ahead with this PR as-is, which will start uploading dev wheels to anaconda.org, then we can test that that works fine, and make sure the wheels work fine, and once we are happy with it, we can remove the Azure dev wheels and update the URL for the dev wheels in the docs and advertise the new link in various places. Once 5.1 is out, we can then also enable the PyPI uploads on stable tags and remove the Azure config for stable builds. So I think this is ready for review - before we merge I will trigger a workflow_dispatch on my fork to make sure all the wheels build fine (since they don't all run as part of a PR).

**Comment by pllim** [comment]
There are unchecked boxes in the post above, so are we merging or not?

**Comment by astrofrog** [comment]
All my TODO checkboxes are now ticked, so this can be reviewed/merged unless anyone objects.

File: .github/workflows/publish.yml
Comment (pllim) [inline .github/workflows/publish.yml:29]: Do we need to add `github.repository == 'astropy/astropy'` here to avoid forks with Actions enabled running this?

https://github.com/astropy/astropy/blob/f0e2129aad460ea0db00730f10f98fe799aefb2f/.github/workflows/ci_cron_weekly.yml#L33

Comment (astrofrog) [reply .github/workflows/publish.yml:29]: Ah yes good catch!

File: astropy/coordinates/tests/test_representation.py
Comment (pllim) [inline astropy/coordinates/tests/test_representation.py:2103]: `assert_allclose_quantity` does not exist. How does this not error out? Am I going crazy?

Comment (pllim) [reply astropy/coordinates/tests/test_representation.py:2103]: Argh nvm... but why?

```python
from astropy.tests.helper import (assert_quantity_allclose as
                                  assert_allclose_quantity)
```

**Comment by pllim** [comment]
In that case, I will approve after you add in checks to ensure this won't run on forks. Thanks!

**Comment by astrofrog** [comment]
@pllim - I've added a condition to make sure it does not run on forks

File: .github/workflows/publish.yml
Comment (astrofrog) [reply .github/workflows/publish.yml:52]: We do for LTS so I was thinking we can backport and then remove these wheels in main

File: .github/workflows/publish.yml
Comment (pllim) [reply .github/workflows/publish.yml:76]: I cannot tell if this comment is supposed to be there or not.

Comment (Cadair) [reply .github/workflows/publish.yml:76]: Now these builds are commented out this comment isn't relevant

Comment (pllim) [reply .github/workflows/publish.yml:76]: Oh yeah, you are right.

**Comment by astrofrog** [comment]
I'll try and wrap this up tomorrow or Thursday - thanks for the review @Cadair!

File: .github/workflows/publish.yml
Comment (astrofrog) [reply .github/workflows/publish.yml:76]: Let me double check that these are still passing first

**Comment by astrofrog** [comment]
The musllinux Python 3.10 wheel seems to take forever to build so I might need to disable that one. I'll let it run to see in the log what is taking a while.

**Comment by astrofrog** [comment]
I've decided to try and debug this, so do not merge :)

**Comment by astrofrog** [comment]
It might have been a fluke, I have restarted the builds

**Comment by astrofrog** [comment]
Yeah it's working now!

File: .github/workflows/publish.yml
Comment (astrofrog) [reply .github/workflows/publish.yml]: Fixed

File: .github/workflows/publish.yml
Comment (astrofrog) [reply .github/workflows/publish.yml]: Done

**Comment by astrofrog** [comment]
@pllim @Cadair - I believe this is ready

**Review by Cadair — APPROVED** [review]

**Review by pllim — COMMENTED** [review]
I still think `github.repository == 'astropy/astropy'` should be applied to all the jobs and not only for scheduled event.

**Comment by pllim** [comment]
@astrofrog , are you able to do a manual backport of this to v5.0.x ? Thanks!

**Pull request merged**
